WordPress/gutenberg
# Fix useZoomOut inserter behavior

Fixes https://github.com/WordPress/gutenberg/issues/66328



## What?

Track if the useZoomOut hook changed the zoom level for the user. If we never changed it, don't change the zoom level on unmount.

Request from @richtabor: If we did change zoom level via the `useZoomOut` hook (controlled mode) _always `resetZoomLevel()` on unmount_.

## Why?

Fixes a bug where zoom out was engaged/disengaged incorrectly on unmount. 

## How?

Adds a ` controlZoomLevelRef`  that tracks if we ever changed the zoom level for the user. When unmounting, if we changed the zoom level, reset it _unless they manually toggled the zoom level after us_. In that condition, `controlZoomLevelRef` is set to `false`.

## Testing Instructions





The key concepts here are:
- If we most recently changed the zoom level for them (in or out), we always resetZoomLevel() level when unmounting.
- If the user most recently changed the zoom level (manually toggling), we do nothing when unmounting.
 
So, with these in mind, we have two scenarios to watch for:
- Closing the inserter always resets the zoom level if the inserter changed the zoom level (clicking on blocks tab from zoomed out state OR clicking on patterns or media tab from zoomed in state)
- Closing the inserter NEVER affects zoom level if zoom state was manually toggled via the zoom out button. However, if you manually toggle, then click to a tab that sets the zoom state, we are now controlling it again.


1. Enter Zoom out manually
2. Open the inserter.
3. The pattern tab should be selected.
4. Exit zoom out manually.
5. Close the inserter.
6. Editor should remain in non-zoomed state (regular editing)

From full screen mode
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Close inserter
5. Zoom out should disengage

From full screen mode, exit zoom out manually
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Manually exit zoom out
5. Close inserter
6. Zoom out should not engage (remain in full screen mode)

## Screen Recording

https://github.com/user-attachments/assets/082ce511-c038-4de4-aa92-5c35f7ef6a5f
